File: React_Express_Fullstack/src/services/authService.ts
import { User, EmailDispatchRecord } from '../types';
import { AUTHORIZED_ADMIN_EMAIL } from '../context/AppContext';
import { API_BASE } from './api';

export interface GoogleAuthResponse {
  success: boolean;
  user?: User;
  requiresOtp?: boolean;
  email?: string;
  name?: string;
  maskedEmail?: string;
  tempSessionId?: string;
  expiresAt?: number;
  devOtpCode?: string;
  error?: string;
  message?: string;
}

export interface VerifyOtpResponse {
  success: boolean;
  user?: User;
  error?: string;
  message?: string;
  attemptsLeft?: number;
}

/**
 * Validates that an email ends strictly with @gmail.com (case-insensitive)
 */
export function isStrictGmail(email: string): boolean {
  if (!email) return false;
  return /^[a-zA-Z0-9._%+-]+@gmail\.com$/i.test(email.trim());
}

/**
 * Masks a gmail address for privacy display (e.g. ss***99@gmail.com)
 */
export function formatMaskedGmail(email: string): string {
  const parts = email.split('@');
  if (parts.length !== 2) return email;
  const user = parts[0];
  const domain = parts[1];
  if (user.length <= 3) {
    return `${user.slice(0, 1)}***@${domain}`;
  }
  return `${user.slice(0, 2)}***${user.slice(-2)}@${domain}`;
}

/**
 * Authenticates via official Google OAuth 2.0 with backend validation
 */
export async function authenticateWithGoogle(params: {
  email: string;
  name?: string;
  avatar?: string;
  googleIdToken?: string;
}): Promise<GoogleAuthResponse> {
  const cleanEmail = (params.email || '').trim().toLowerCase();

  // Strict domain verification on the client
  if (!isStrictGmail(cleanEmail)) {
    return {
      success: false,
      error: 'ACCESS_DENIED_NON_GMAIL',
      message: `Access Denied: Only official Google accounts ending in @gmail.com are permitted on AdamFlix. The domain "${cleanEmail.split('@')[1] || 'non-gmail'}" is completely blocked.`
    };
  }

  try {
    const res = await fetch(`${API_BASE}/api/auth/google/verify`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(params)
    });

    const data = await res.json();
    if (!res.ok) {
      return {
        success: false,
        error: data.error || 'AUTH_FAILED',
        message: data.message || 'Authentication failed. Please verify your Google credentials.'
      };
    }
    return data;
  } catch (err) {
    console.warn('[authService] Backend API offline, executing standalone fallback for preview environment:', err);
    // Standalone fallback: simulate OTP requirement for non-admin new logins
    const isSuperAdmin = cleanEmail === AUTHORIZED_ADMIN_EMAIL.toLowerCase();
    
    // For test evaluation, issue a 6-digit code
    const devCode = Math.floor(100000 + Math.random() * 900000).toString();
    const tempSession = `sess_${Date.now()}`;
    
    // Save to local storage for verification fallback
    sessionStorage.setItem(`otp_${cleanEmail}`, JSON.stringify({
      code: devCode,
      name: params.name || (isSuperAdmin ? 'Souhaimat (Administrator)' : cleanEmail.split('@')[0]),
      avatar: params.avatar || 'https://images.unsplash.com/photo-1534528741775-53994a69daeb?w=200&auto=format&fit=crop&q=80',
      expiresAt: Date.now() + 10 * 60 * 1000
    }));

    return {
      success: true,
      requiresOtp: true,
      email: cleanEmail,
      name: params.name || cleanEmail.split('@')[0],
      maskedEmail: formatMaskedGmail(cleanEmail),
      tempSessionId: tempSession,
      devOtpCode: devCode,
      expiresAt: Date.now() + 10 * 60 * 1000,
      message: `A 6-digit verification code has been dispatched to ${cleanEmail}.`
    };
  }
}

/**
 * Verifies the 6-digit OTP code with the backend
 */
export async function verifyOtpCode(params: {
  email: string;
  code: string;
  tempSessionId?: string;
}): Promise<VerifyOtpResponse> {
  const cleanEmail = (params.email || '').trim().toLowerCase();
  const cleanCode = (params.code || '').trim();

  try {
    const res = await fetch(`${API_BASE}/api/auth/otp/verify`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        email: cleanEmail,
        code: cleanCode,
        tempSessionId: params.tempSessionId
      })
    });

    const data = await res.json();
    if (!res.ok) {
      return {
        success: false,
        error: data.error || 'INVALID_CODE',
        message: data.message || 'Incorrect verification code. Please check your Gmail.',
        attemptsLeft: data.attemptsLeft
      };
    }
    return data;
  } catch (err) {
    console.warn('[authService] Backend offline, checking sessionStorage fallback:', err);
    const raw = sessionStorage.getItem(`otp_${cleanEmail}`);
    if (raw) {
      try {
        const stored = JSON.parse(raw);
        if (stored.code === cleanCode) {
          const isSuperAdmin = cleanEmail === AUTHORIZED_ADMIN_EMAIL.toLowerCase();
          const user: User = {
            id: isSuperAdmin ? 'u-admin-101' : `u-${Date.now()}`,
            name: stored.name,
            email: cleanEmail,
            avatar: stored.avatar,
            role: isSuperAdmin ? 'admin' : 'user',
            joinedDate: new Intl.DateTimeFormat('en-US', { month: 'long', year: 'numeric' }).format(new Date()),
            isVerified: true,
            authProvider: 'google'
          };
          sessionStorage.removeItem(`otp_${cleanEmail}`);
          return { success: true, user, message: 'Account activated successfully!' };
        }
      } catch {
        // continue to error
      }
    }
    return {
      success: false,
      error: 'INVALID_CODE',
      message: 'Invalid verification code. Please try again.'
    };
  }
}

/**
 * Resends a fresh 6-digit code
 */
export async function resendOtpCode(email: string): Promise<{ success: boolean; devOtpCode?: string; message?: string }> {
  const cleanEmail = (email || '').trim().toLowerCase();

  try {
    const res = await fetch(`${API_BASE}/api/auth/otp/resend`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ email: cleanEmail })
    });
    const data = await res.json();
    return data;
  } catch {
    const newCode = Math.floor(100000 + Math.random() * 900000).toString();
    const existing = sessionStorage.getItem(`otp_${cleanEmail}`);
    let storedName = cleanEmail.split('@')[0];
    let storedAvatar = 'https://images.unsplash.com/photo-1534528741775-53994a69daeb?w=200&auto=format&fit=crop&q=80';
    if (existing) {
      try {
        const parsed = JSON.parse(existing);
        storedName = parsed.name || storedName;
        storedAvatar = parsed.avatar || storedAvatar;
      } catch {}
    }
    sessionStorage.setItem(`otp_${cleanEmail}`, JSON.stringify({
      code: newCode,
      name: storedName,
      avatar: storedAvatar,
      expiresAt: Date.now() + 10 * 60 * 1000
    }));
    return { success: true, devOtpCode: newCode, message: `New code sent to ${cleanEmail}` };
  }
}

/**
 * Fetches the latest email record from server for inspection
 */
export async function fetchDispatchedEmail(email: string): Promise<EmailDispatchRecord | null> {
  try {
    const res = await fetch(`${API_BASE}/api/auth/otp/inspect?email=${encodeURIComponent(email)}`);
    if (!res.ok) return null;
    const data = await res.json();
    return data.record || null;
  } catch {
    return null;
  }
}

File: Services/AuthService.cs
using System.Collections.Concurrent;
using System.Security.Cryptography;
using AdamFlix.Models;

namespace AdamFlix.Services
{
    public interface IAuthService
    {
        User? GetUserByEmail(string email);
        GeneratedOtp GenerateOtpForUser(string email, string? name = null, string? avatar = null);
        OtpVerificationResult VerifyOtp(string email, string code);
        string MaskEmail(string email);
    }

    public class AuthService : IAuthService
    {
        public const string AUTHORIZED_ADMIN_EMAIL = "ssouhaimat1999@gmail.com";

        private class PendingOtp
        {
            public string Email { get; set; } = string.Empty;
            public string Name { get; set; } = string.Empty;
            public string Avatar { get; set; } = string.Empty;
            public string Code { get; set; } = string.Empty;
            public long ExpiresAt { get; set; }
            public int AttemptsLeft { get; set; } = 5;
            public string TempSessionId { get; set; } = string.Empty;
        }

        private readonly ConcurrentDictionary<string, User> _users = new();
        private readonly ConcurrentDictionary<string, PendingOtp> _pendingOtps = new();

        public AuthService()
        {
            // Seed verified super admin
            _users.TryAdd(AUTHORIZED_ADMIN_EMAIL.ToLowerInvariant(), new User
            {
                Id = "u-admin-101",
                Name = "Souhaimat (Administrator)",
                Email = AUTHORIZED_ADMIN_EMAIL,
                Avatar = "https://images.unsplash.com/photo-1534528741775-53994a69daeb?w=200&auto=format&fit=crop&q=80",
                Role = "admin",
                IsVerified = true,
                AuthProvider = "google"
            });
        }

        public User? GetUserByEmail(string email)
        {
            _users.TryGetValue(email.ToLowerInvariant(), out var user);
            return user;
        }

        public GeneratedOtp GenerateOtpForUser(string email, string? name = null, string? avatar = null)
        {
            var cleanEmail = email.ToLowerInvariant();
            var code = RandomNumberGenerator.GetInt32(100000, 999999).ToString();
            var sessionId = $"sess_{DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()}";
            var expiresAt = DateTimeOffset.UtcNow.AddMinutes(10).ToUnixTimeMilliseconds();

            var pending = new PendingOtp
            {
                Email = cleanEmail,
                Name = name ?? cleanEmail.Split('@')[0],
                Avatar = avatar ?? "https://images.unsplash.com/photo-1534528741775-53994a69daeb?w=200&auto=format&fit=crop&q=80",
                Code = code,
                ExpiresAt = expiresAt,
                AttemptsLeft = 5,
                TempSessionId = sessionId
            };

            _pendingOtps.AddOrUpdate(cleanEmail, pending, (_, _) => pending);

            return new GeneratedOtp
            {
                Code = code,
                TempSessionId = sessionId,
                ExpiresAt = expiresAt
            };
        }

        public OtpVerificationResult VerifyOtp(string email, string code)
        {
            var cleanEmail = email.ToLowerInvariant();

            if (!_pendingOtps.TryGetValue(cleanEmail, out var pending))
            {
                return new OtpVerificationResult
                {
                    Success = false,
                    ErrorCode = "NO_PENDING_OTP",
                    Message = "No pending verification found. Please sign in again."
                };
            }

            if (DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() > pending.ExpiresAt)
            {
                _pendingOtps.TryRemove(cleanEmail, out _);
                return new OtpVerificationResult
                {
                    Success = false,
                    ErrorCode = "OTP_EXPIRED",
                    Message = "This verification code has expired. Please request a new one."
                };
            }

            if (pending.AttemptsLeft <= 0)
            {
                _pendingOtps.TryRemove(cleanEmail, out _);
                return new OtpVerificationResult
                {
                    Success = false,
                    ErrorCode = "TOO_MANY_ATTEMPTS",
                    Message = "Too many incorrect attempts. Please request a new code."
                };
            }

            if (pending.Code != code.Trim())
            {
                pending.AttemptsLeft -= 1;
                return new OtpVerificationResult
                {
                    Success = false,
                    ErrorCode = "INVALID_OTP",
                    AttemptsLeft = pending.AttemptsLeft,
                    Message = $"Invalid code. {pending.AttemptsLeft} attempt(s) remaining."
                };
            }

            var isAdmin = cleanEmail.Equals(AUTHORIZED_ADMIN_EMAIL, StringComparison.OrdinalIgnoreCase);
            var activatedUser = new User
            {
                Id = isAdmin ? "u-admin-101" : $"u-{DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()}",
                Name = pending.Name,
                Email = cleanEmail,
                Avatar = pending.Avatar,
                Role = isAdmin ? "admin" : "user",
                IsVerified = true,
                AuthProvider = "google"
            };

            _users.AddOrUpdate(cleanEmail, activatedUser, (_, _) => activatedUser);
            _pendingOtps.TryRemove(cleanEmail, out _);

            return new OtpVerificationResult
            {
                Success = true,
                User = activatedUser,
                Message = "Account verified successfully."
            };
        }

        public string MaskEmail(string email)
        {
            var parts = email.Split('@');
            if (parts.Length != 2) return email;
            var u = parts[0];
            return u.Length <= 3 ? $"{u[0]}***@{parts[1]}" : $"{u[..2]}***{u[^2..]}@{parts[1]}";
        }
    }
}
